src/magicbox/ResultPreviewsGrid.ts

```diff
@@ -0,0 +1,268 @@
+    this.root = Component.resolveRoot(parentContainer);
+    this.options = defaults(options, <IResultPreviewsGridOptions>{
+      selectedClass: 'magic-box-selected'
+    });
+    this.buildResultsPreviewsContainer();
+  }
+
+  public isHoverKeyboardControlled() {
+    return this.keyboardSelectionMode;
+  }
+
+  public setDisplayedItems(items: ISearchResultPreview[]) {
+    this.clearDisplayedItems();
+    items.forEach(preview =>
+      this.appendSearchResultPreview({
+        element: preview.element.cloneNode(true) as HTMLElement,
+        onSelect: preview.onSelect
+      })
+    );
+  }
+
+  public getHoveredItem(): ISearchResultPreview {
+    if (this.activePreviews.length === 0) {
+      return null;
+    }
+    const previewId = this.getHoveredItemId();
+    return this.activePreviews[previewId] || null;
+  }
+
+  public tryHoverOnFirstItem() {
+    if (this.activePreviews.length === 0) {
+      return (this.keyboardSelectionMode = false);
+    }
+    this.setHoveredItemFromId(0);
+    return (this.keyboardSelectionMode = true);
+  }
+
+  public tryHoverOnNextItem(direction: ResultPreviewsGridDirection) {
+    const currentSelectionId = this.getHoveredItemId();
+    if (currentSelectionId === null) {
+      return false;
+    }
+    const totalLength = this.activePreviews.length;
+    const rowLength = this.calculatePreviewsPerRow();
+    switch (direction) {
+      case ResultPreviewsGridDirection.Left:
+        if (currentSelectionId === 0) {
+          return false;
+        }
+        if (currentSelectionId % rowLength === 0) {
+          return false;
+        }
+        this.setHoveredItemFromId(currentSelectionId - 1);
+        this.keyboardSelectionMode = true;
+        return true;
+      case ResultPreviewsGridDirection.Right:
+        if (currentSelectionId === totalLength - 1) {
+          return false;
+        }
+        if (currentSelectionId % rowLength === rowLength - 1) {
+          return false;
+        }
+        this.setHoveredItemFromId(currentSelectionId + 1);
+        this.keyboardSelectionMode = true;
+        return true;
+      case ResultPreviewsGridDirection.Up:
+        if (currentSelectionId - rowLength < 0) {
+          return false;
+        }
+        this.setHoveredItemFromId(currentSelectionId - rowLength);
+        this.keyboardSelectionMode = true;
+        return true;
+      case ResultPreviewsGridDirection.Down:
+        if (currentSelectionId + rowLength >= totalLength) {
+          return false;
+        }
+        this.setHoveredItemFromId(currentSelectionId + rowLength);
+        this.keyboardSelectionMode = true;
+        return true;
+    }
+  }
+
+  public clearHover() {
+    const currentSelection = this.getHoveredItemElement();
+    if (!currentSelection) {
+      return;
+    }
+    this.deselectElement(currentSelection);
+    $$(this.root).trigger(ResultPreviewsGridEvents.PreviewHovered, <IPreviewHoveredEventArgs>{
+      preview: null
+    });
+  }
+
+  public keyboardSelect() {
```

Comment:
Currently, `SuggestionsManager` is responsible for handling keyboard events (arrow keys and enter, specifically) while this component only handles mouse events. Dedicating the enter key event to `ResultPreviewsGrid` would be possible, but may be inconsistent with the rest of `ResultPreviewsGrid`'s behaviour since other keys would still be handled by `SuggestionsManager`.

Unfortunately, handling all key events in `ResultPreviewsGrid` for consistency would make it difficult for the it to ignore up/down keys when the focus is on another component like `SuggestionsList` and would make it difficult for `SuggestionsManager` to know when to switch the focus back to `SuggestionsList`.